dev/buildtool/build_commands.py

```diff
@@ -32,15 +33,101 @@
     check_subprocess,
+    check_subprocesses_to_logfile,
     determine_logfile_path,
+    ensure_dir_exists,
     timedelta_string,
     timestring,
-    check_subprocesses_to_logfile,
     write_to_path)
 
 
+class GradleMetricsUpdater(object):
+  """Collect gradle metrics, focusing on characterizing failures."""
+
+  def __init__(
+      self, metrics_repository, spinnaker_repository, gradle_context_name):
+    self.__metrics = metrics_repository
+    self.__name = spinnaker_repository.name
+    self.__context = gradle_context_name
+
+  def __call__(self, retcode, output):
+    """Update metrics considering the final return code and process output."""
+    labels = self.determine_labels(retcode, output)
+    self.__metrics.register_counter(
+        'GradleOutcome', 'Outcomes when running gradle', labels.keys())
+    counter = self.__metrics.counter('GradleOutcome', labels)
+    counter.inc()
+    return counter
+
+  def extract_failure_summary(self, retcode, output):
+    """Determine the top level failure message from gradle."""
+    summary_match = re.search(
+        "Execution failed for task '(.+)'.*\n", output, re.MULTILINE)
+    if not retcode or not summary_match:
+      return None, None
+
+    failed_task = summary_match.group(1)
+    remainder = output[summary_match.end(0):]
+    first_line = remainder[:remainder.find('\n')]
+    logging.debug('Instrumenting failure from after stdout %d - %d: %s',
+                  summary_match.start(0), summary_match.end(0), first_line)
+    return failed_task, first_line
+
+  def determine_labels(self, retcode, output):
+    """Determine label bindings for this outcome."""
+    labels = {
+        'repository': self.__name,
+        'context': self.__context,
+        'success': retcode == 0,
+        'failed_task': '',
+        'failed_by': '',
+        'failed_reason': ''
+    }
+    if retcode == 0:
+      return labels
+
+    failed_task, summary_line = self.extract_failure_summary(retcode, output)
+    self.update_failure_cause(labels, failed_task, summary_line)
+
+    return labels
+
+  def update_failure_cause(self, labels, failed_task, summary_line):
+    """Figure out the labels from the failure summary line."""
+
+    # This is an example of an error (wrapped from one to multiple lines
+    # for readability here. The ...jar was a long path to the bintray jar.
+    #
+    # > Could not upload to \
+    # 'https://api.bintray.com/content/...jar': HTTP/1.1 409 Conflict \
+    # [message:Unable to upload files: An artifact with the path \
+    # 'com/netflix/spinnaker/echo/echo-core/1.542.0/echo-core-1.542.0.jar' \
+    # already exists]
+
+    summary_match = re.match(
+        r"[> ]*(.*?)'(.+?)': HTTP/[0-9\.]+ ([0-9]{3}) ([\w]+)(.*)",
+        summary_line)
+    failed_by = 'unknown'
+    failed_reason = 'unknown'
+    if not summary_match:
+      logging.error('Cannot interpret cause of failure:\n%s', summary_line)
+    else:
+      # abstract = summary_match.group(1)
```

Comment:
Will be used in the future?

Reply:
i dont know. it's there for documentation. As I encounter and handle more error cases I might change the label schema to better accomodate them. For now we just want to be able to characterize and quantify impediments to building artifacts.